rpc/json-rpc/src/response.rs

```diff
@@ -0,0 +1,434 @@
+    where
+        S: Serializer,
+    {
+        let mut s = serializer.serialize_struct("Response", 3)?;
+
+        s.serialize_field("jsonrpc", &self.jsonrpc)?;
+
+        // This member is required.
+        //
+        // Even if `null`, or the client `Request` didn't include one.
+        s.serialize_field("id", &self.id)?;
+
+        match &self.payload {
+            Ok(r) => s.serialize_field("result", r)?,
+            Err(e) => s.serialize_field("error", e)?,
+        }
+
+        s.end()
+    }
+}
+
+// [`Response`] has a manual deserialization implementation because
+// we need to confirm `result` and `error` don't both exist:
+//
+// > Either the result member or error member MUST be included, but both members MUST NOT be included.
+//
+// <https://www.jsonrpc.org/specification#error_object>
+impl<'de, T> Deserialize<'de> for Response<T>
+where
+    T: Deserialize<'de> + 'de,
+{
+    fn deserialize<D: Deserializer<'de>>(der: D) -> Result<Self, D::Error> {
+        use std::marker::PhantomData;
+
+        use serde::de::{Error, MapAccess, Visitor};
+
+        /// This type represents the key values within [`Response`].
+        enum Key {
+            /// "jsonrpc" field.
+            JsonRpc,
+            /// "result" field.
+            Result,
+            /// "error" field.
+            Error,
+            /// "id" field.
+            Id,
+            /// Any other unknown field (ignored).
+            Unknown,
+        }
+
+        // Deserialization for [`Response`]'s key fields.
+        //
+        // This ignores unknown keys.
+        impl<'de> Deserialize<'de> for Key {
+            fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+            where
+                D: Deserializer<'de>,
+            {
+                /// Serde visitor for [`Response`]'s key fields.
+                struct KeyVisitor;
+
+                impl Visitor<'_> for KeyVisitor {
+                    type Value = Key;
+
+                    fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+                        formatter.write_str("`jsonrpc`, `result`, `error`, `id`")
+                    }
+
+                    fn visit_str<E>(self, string: &str) -> Result<Key, E>
+                    where
+                        E: Error,
+                    {
+                        match string {
+                            "jsonrpc" => Ok(Key::JsonRpc),
+                            "id" => Ok(Key::Id),
+                            "result" => Ok(Key::Result),
+                            "error" => Ok(Key::Error),
+                            // Ignore any other keys that appear
+                            // and continue deserialization.
+                            _ => Ok(Key::Unknown),
+                        }
+                    }
+                }
+
+                deserializer.deserialize_identifier(KeyVisitor)
+            }
+        }
```

Comment:
Would it be possible to just use `&str` as the key type so we have to do this?

Reply:
I just treid this and it turns out you can't always borrow a str from the data so instead of `&std` `Cow<str>` works

Reply:
discussed in a meeting: https://github.com/monero-project/meta/issues/1021 this is slightly less optimal